### DISCUSSION TITLE
Unit test acceptace regarding datasource rest connector

### DISCUSSION THREAD

**Discussion by rabdev89** [loopbackio/loopback-next#7454] [Q&A]
Hello guys, How to test if the datasource rest link is down? I have created a datasource like strapi api. i wanna put on my unit test it strapi api is down.

and since the endpoint is can access only when user is login needed to be authenticated how to bypass authentication. the authentication is coming from another 3rd party api. what the lb4 application is getting the Bearer token and validate it.
i already tried to use invoke but of course since in my controller i injected authentication it gives me

```
 invokes GET /accounts:
Error: expected 200 "OK", got 401 "Unauthorized"
```

here's my sample code
```

describe('AccountController (acceptance)', () => {
  let app: Application;
  let client: Client;
  const customerNumber: string = '';
  const page: number = 1
  const limit: number = 10;
  const sortBy: string = '';
  const order: string = '';
  const status: string = '';
  const token: string = '';
  before('setupApplication', async () => {
    ({app, client} = await setupApplication());
  });
  after(async () => {
    await app.stop();
  });
  it('GET /accounts', async () => {
    const accountsService = await app.get<AccountsService>('services.AccountsService');
    //this is coming from datasource service
    const alfaAccountService = await app.get<AlfaAccounts>('services.AlfaAccounts');
    const data = await alfaAccountService.getAccounts(customerNumber, page, limit, sortBy, order, status);
    const res = await accountsService.getAllAccounts(data);
    // const res = await client.get('/accounts')
    //     .set('Bearer', token)
    //     .expect(200);
    expect(res).to.containEql(res);
  });
  it('it should return 401 if user is not loggedin', async () => {
    const res = await client.get('/accounts').expect(401);
    expect(res).to.containEql(res);
  });
});
```

**Comment by marioestradarosa** [comment]
For this third party service I guess they have a healthCheck endPoint, which you can place it in a helper function to get the status .
For the authentication, you can also create a helper function (in a separate file , like a fixture) and invoke it in the before to get the token, you can use axios package for that. Once you get the token you can pass it over on subsequent requests like so :

```ts
const res = await client.get('/accounts')
            .set('Authorization', 'Bearer ' + token)
        .expect(200);
```
        
And the helper function inside its own file can be like:

```ts
import axios from 'axios';
export async function givenTokenFromStrapiDirectly () { 
const { data } = await axios.post('http://localhost:1337/auth/local', {
  identifier: 'youridentifier',
  password: 'yourpassword',
});
return data.token;
}
```

The givenTokenFromStrapiDirectly should be invoked like so:

```ts
 before('setupApplication', async () => {
    ({app, client} = await setupApplication());
    token = await givenTokenFromStrapiDirectly(); 
  });
  
  ```
  
Well, that's more or less the logic. Also I encourage you to read the loopback tests and the shopping example application to learn more about how to leverage the test suite at loopback4.

**Reply by rabdev89** [reply]
this is good, but i think this will not work in our current setup. since the application im building is a multi tenant.. we have api for getting the users info.  api for content. For auth fe is the one dealing with it and just throw the Bearer token and they are usng oauth.
I am also new in developing testing.
what im dealing with right now is (api-users-info). it returns html when the server is down (coz there is a certain time that it is up for development purposes). so how to do it in testing? do i  need to create another it() like this.
```
it('it should check if server is down', async () => {
  const res = await client.get('/accounts').expect(500);
  expect(res).to.containEql(res);
});
```
therefore, in every endpoint or describe() i need add somehing like this.

For auth,  maybe i need to include the creation of token for using RSA SHA256 and validate
and pass it.
